streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -91,10 +91,14 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final Intern
         try {
             context.setRecordContext(entry.recordContext());
             if (flushListener != null) {
-                final V oldValue = sendOldValues ? serdes.valueFrom(underlying.get(entry.key())) : null;
+                V oldValue = null;
+                if (sendOldValues) {
+                    final byte[] oldBytesValue = underlying.get(entry.key());
+                    oldValue = oldBytesValue == null ? null : serdes.valueFrom(oldBytesValue);
+                }
                 underlying.put(entry.key(), entry.newValue());
                 flushListener.apply(serdes.keyFrom(entry.key().get()),
-                                    serdes.valueFrom(entry.newValue()),
+                                    entry.newValue() == null ? null : serdes.valueFrom(entry.newValue()),
```

Comment:
Wondering if we need this check. It's of course safer, but the contract should be, that the Serde return `null` on `null` ? Thus, it would be a user error? (Same for all other places where the check is added.) Not saying we should have have it, just want to double check about the reasoning.

Reply:
We actually talked about this on the JIRA and I thought you're preferring this option :) actually, that may not be true. What we proposed is:

```
1) Clarify in javadoc that null value in put(key, value) indicates deletion; if it is "null" object by-pass the serde and send "null" bytes directly into inner functions and vice verse for deserialization; do not enforce user customized serdes how to handle null values since we are not going to call them with null values any more.

2) Do NOT clarify in java doc that null value in put(key, value) indicates deletion; implement all delete(key) call directly throughout all the layers of stores instead of calling put(key, null); recommend user customized serdes to handle null values themselves.
```

I'm following option 1) here, and I think what you suggested here is a bit different with both of them:

```
3) Clarify that any user customized serde must map null to null; if a user failed to do so it is her problem.
```

I personally am inclined to 1) over 3).

Reply:
My inclination is to go with option 2 as we discussed in the Jira, but if it's between option 1 and 3 presented here then I'll take 1 over 3 as well

Reply:
I was not sure if we did agree on anything... Also just reread the discussion. I don't think that option (2) works, because we inherit delete semantics from the broker compaction and thus, cannot avoid it.

The question is really if you put the burden on the user to implement a correct Serde (that must map `null` to `null`), or if we do the check within the library (safer, but also some runtime overhead...).

If we go with (1) should we add checks that serializer does never return `null` ?

Reply:
About 2), yes, we dropped this option exactly because kafka side log compaction does treat `null` value specially anyways, and hence we decided to go with 1) (at least that was my understanding..) I'm only copying the content for context in the previous reply, and assuming that we have selected 1) over 2).

And now about 1) v.s. 3), like I said I think 3) is less efficient since it requires everyone to implement `null` to `null` mapping which is too much, so I'm still inclined to 1).

As for implementation, even if user mapped `not-null` to `null` in ser, the most inner stores (memory map and rocks) would still treat `null` bytes specially as `delete`; as for deser, if the user maps `not-null` bytes to `null` objects, then we would just return that in the api call and that should be fine since it is user's call to interpret `null` object as "not exist" or "a dummy value" then.

Reply:
Yes. So if users have buggy ser code and return `null` for non-`null` input, they effectively delete data accidentally. Should we add a note to Serde/Serializer JavaDocs (web docs)? As an alternative, we should even check if ser returns `null` for non-`null` input and throw error to notify users about their buggy ser code?

I am fine with going with (1) -- might be better to protect users from potentially buggy serde code. Just wanted to clarify our decision.